daemon/smartcard.c

```diff
@@ -331,6 +584,137 @@ smartcard_cb_generic(int fd, unsigned events, event_io_t *io, void *data)
+			return;
+		}
+		switch (msg->code) {
+		case TOKEN_TO_DAEMON__CODE__CHANGE_PIN_SUCCESSFUL: {
+			control_send_message(CONTROL_RESPONSE_CONTAINER_CHANGE_PIN_SUCCESSFUL,
+					     resp_fd);
+		} break;
+		case TOKEN_TO_DAEMON__CODE__CHANGE_PIN_FAILED: {
+			control_send_message(CONTROL_RESPONSE_CONTAINER_CHANGE_PIN_FAILED, resp_fd);
+		} break;
+		case TOKEN_TO_DAEMON__CODE__PROVISION_PIN_SUCCESSFUL: {
+			char *path =
+				mem_printf("%s/%s", container_get_images_dir(startdata->container),
+					   TOKEN_IS_INIT_FILE_NAME);
+			rc = file_touch(path);
+			if (rc != 0) {
+				ERROR("Could not write file %s to flag that container %s's token has been initialized\n \
+						This may leave the system in an inconsistent state!",
+				      path, uuid_string(container_get_uuid(startdata->container)));
+				container_set_token_is_linked_to_device(startdata->container,
+									false);
+				control_send_message(CONTROL_RESPONSE_CONTAINER_CHANGE_PIN_FAILED,
+						     resp_fd);
+			} else {
+				container_set_token_is_linked_to_device(startdata->container, true);
+				control_send_message(
+					CONTROL_RESPONSE_CONTAINER_CHANGE_PIN_SUCCESSFUL, resp_fd);
+			}
+		} break;
+		case TOKEN_TO_DAEMON__CODE__PROVISION_PIN_FAILED: {
+			container_set_token_is_linked_to_device(startdata->container, false);
+			control_send_message(CONTROL_RESPONSE_CONTAINER_CHANGE_PIN_FAILED, resp_fd);
+		} break;
+		default:
+			ERROR("TokenToDaemon command %d not expected as answer to change_pin",
+			      msg->code);
+		}
+		control_send_message(CONTROL_RESPONSE_CONTAINER_CHANGE_PIN_FAILED, resp_fd);
+		protobuf_free_message((ProtobufCMessage *)msg);
+		event_remove_io(io);
+		event_io_free(io);
+		mem_free(startdata);
+	}
+}
+
+int
+smartcard_change_container_pin(smartcard_t *smartcard, control_t *control, container_t *container,
+			       const char *passwd, const char *newpasswd)
+{
+	ASSERT(smartcard);
+	ASSERT(container);
+	ASSERT(control);
+	ASSERT(passwd);
+	ASSERT(newpasswd);
+
+	int ret = -1;
+	int pw_size = strlen(passwd);
+	int newpw_size = strlen(newpasswd);
+	unsigned char pair_sec[MAX_PAIR_SEC_LEN];
+	int resp_fd = control_get_client_sock(control);
+	bool is_provisioning;
+
+	smartcard_startdata_t *startdata = mem_alloc(sizeof(smartcard_startdata_t));
+	if (!startdata) {
+		ERROR("Could not allocate memory for startdata");
+		return -1;
+	}
+	startdata->smartcard = smartcard;
+	startdata->container = container;
+	startdata->control = control;
+
+	DEBUG("SCD: Passwd form UI: %s, size: %d", passwd, pw_size);
+	DEBUG("SCD: New Passwd form UI: %s, size: %d", newpasswd, newpw_size);
+
+	ret = smartcard_get_pairing_secret(smartcard, pair_sec, sizeof(pair_sec));
+	if (ret < 0) {
+		ERROR("Could not retrieve pairing secret, ret code : %d", ret);
+		control_send_message(CONTROL_RESPONSE_CONTAINER_CHANGE_PIN_FAILED, resp_fd);
+		return -1;
+	}
+
+	is_provisioning = !smartcard_container_token_is_provisioned(container);
```

Comment:
use same datatype that function returns (int ?)

Reply:
i don't understand, smartcard_container_token_is_provisioned() returns a bool